services/skus/receipt.go

```diff
@@ -62,103 +64,113 @@ func (v *receiptVerifier) validateApple(ctx context.Context, req model.ReceiptRe
 
 	resp := &appstore.IAPResponse{}
 	if err := v.appStoreCl.Verify(ctx, asreq, resp); err != nil {
-		return "", fmt.Errorf("failed to verify receipt: %w", err)
+		return model.ReceiptData{}, fmt.Errorf("failed to verify receipt: %w", err)
 	}
 
 	// ProductID on an InApp object must match the SubscriptionID.
 	//
 	// By doing so we:
 	// - find the purchase that is being verified (i.e. to disambiguate VPN from Leo);
 	// - utilise Apple verification to make sure the client supplied data (SubscriptionID) is valid and to be trusted.
-	item, ok := findInAppBySubID(resp.Receipt.InApp, req.SubscriptionID)
+	item, ok := findInAppBySubID(resp.Receipt.InApp, req.SubscriptionID, now)
 	if ok {
-		return item.OriginalTransactionID, nil
+		return newReceiptDataApple(req, item), nil
 	}
 
 	// Try finding in latest_receipt_info.
-	item, ok = findInAppBySubID(resp.LatestReceiptInfo, req.SubscriptionID)
+	item, ok = findInAppBySubID(resp.LatestReceiptInfo, req.SubscriptionID, now)
 	if ok {
-		return item.OriginalTransactionID, nil
+		return newReceiptDataApple(req, item), nil
 	}
 
 	// Special case for VPN.
 	// The client may send bravevpn.monthly as subscription_id for bravevpn.yearly product.
 	if req.SubscriptionID == "bravevpn.monthly" {
-		item, ok := findInAppBySubID(resp.Receipt.InApp, "bravevpn.yearly")
+		item, ok := findInAppBySubID(resp.Receipt.InApp, "bravevpn.yearly", now)
 		if ok {
-			return item.OriginalTransactionID, nil
+			return newReceiptDataApple(req, item), nil
 		}
 
-		item, ok = findInAppBySubID(resp.LatestReceiptInfo, "bravevpn.yearly")
+		item, ok = findInAppBySubID(resp.LatestReceiptInfo, "bravevpn.yearly", now)
 		if ok {
-			return item.OriginalTransactionID, nil
+			return newReceiptDataApple(req, item), nil
```

Comment:
question: is it okay to drop through to the [legacy check](https://github.com/brave-intl/bat-go/pull/2587/files#diff-6ebc9187728cc8ddacabb1ef3b11220d2fab15950be2e72861dd9e93ee171268R102) if non of the two conditions above are `ok`?

Reply:
Good question, and the answer is yes. While I could have handled it right here, if none of the conditions has worked, then it'll be cut off in the legacy check you've linked.

I did that intentionally with two thoughts in mind:
- to reduce the total number of error handling statements – there's only one for the amount of the branches, which is good;
- it might be necessary to have some additional logic down in the legacy path (though unlikely).

Like, the legacy check is the final judge, and if nothing could be done, then error out.